### ISSUE TITLE
In-memory caching

### ISSUE THREAD

**Issue by lars-t-hansen** [NAICNO/Jobanalyzer#506]
Parsed data should be cached in memory (and eventually there may be indexes to speed up finding data, but not until this need is clear).  The purpose of this is to allow a large volume of queries to be served quickly.  We envision that the REST API will be (ab)used by many simple queries rather than by larger optimized queries.

Roughly we will cache lightly postprocessed data (corrected for gpu memory, having filtered duplicate timestamps, other context-insensitive things) per-file.  Fields whose values depend on the exact stream and may thus be affected by a partial stream may be moved out of the Sample structure and computed on-demand in a data structure on the side.  For all practical purposes, the cached data will be considered read-only once read.

When code requests data using the existing API, cached data will be stiched together into the final sample stream and duplicate timestamps at the boundary will be removed then and filtering applied.  Apart from the fields that move out of Sample, client code will notice very little difference.  The `parse` command will be somewhat affected since the raw data will not be available, only the filtered streams (that have to be stiched together).  So maybe `parse` in its raw mode gets to use a lower-level API that always rereads the data without postprocessing, this is probably OK.

A cluster config file may become mandatory to all operations to allow caching to be sane.

Caching will be controlled by a `-cache` switch to `sonalyze daemon` that specifies the size of the cache.  The limit is soft: data that are held by an ongoing query can be evicted from the cache but won't be removed from memory until the query is over and the GC runs.

A rough plan:

- [x] Change code to require a config file for anything that reads log data
- [x] Change the API to SampleCluster.ReadSamples so that it delivers InputStreamSets, this moves the invocation of postprocessing into sonarlog for now and changes all clients
- [x] Make ReadSamples also compute and return the bounds for those data, since most clients need them and it's not costly to compute them, especially once the streams have been sorted before filtering, this hides ComputeTimeBounds from all clients
- [x] Move the invocation of computation of cpu_util_pct and filtering and packing of the array out of the normal postprocessing, it will show up where the invocation of postprocessing is now.  This will still update the streams in-place.
- [x] Split the cpu_util_pct field out from Sample data and (probably) move the invocation of the computation to places that need the data.  This looks like `jobs`, `load`, `profile` exclusively.

At this point, it should become possible to cache the postprocessed data per-file, and then stich those streams together when requested, removing any duplicate timestamps at the boundary.  Some high-level points:

- [x] figure out a cache structure, capacity, eviction etc
- [x] actually cache the data and figure out stiching

**Comment by lars-t-hansen** [comment]
Things aren't 100% structured as above but this is code complete, unit testing is done, and integration/deployment testing is ongoing.  For job queries against 30 days of Saga data, with a suitably large cache, this drops query time by half - the time it took to do I/O before.  Now it'll be interesting to look at where the rest of the time goes, but that's not part of the work here.
